### ISSUE TITLE
Packet loss when establishing TRUN connection

### ISSUE THREAD

**Issue by staowyze** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#185]
Hi, I am using TURN over UDP, and experienced issues within unreliable network.
e.g. if a create permission packet is sent, but the success response packet is dropped by wireless network, the connection would fail without retrying, which seems not ideal.
Another issue would be, we would report TURN relay address once allocation is done
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Ice/TurnConnection.c#L892
however, the connection is not ready yet, failure could happen while creating permission and binding.

**Comment by zhiyua-git** [comment]
Hi @staowyze ,

Thanks for reporting, will fix the issue soon.

**Comment by chehefen** [comment]
Hi @staowyze, do you mind share with us how you simulate unreliable network. It would be easier for us to diagnose the issue if we can repro your situation. Right now turn create permission packet does get resent periodically within a timeout https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/206eb882130a6d9cfa6bf401d315accf34cc0df0/src/source/Ice/TurnConnection.c#L1133. Regarding your other issue. it's ok if the connection is not ready yet, if the connection establishment failures then turn won't pass the ice connection check and won't get selected anyways.

**Comment by staowyze** [comment]
do you know how long would it retry? From what I have seen, it would stuck forever.
TURN relay address is reported right after allocation, do you think it's better to claim after binding?
I didn't simulate unreliable network, my wireless is not stable because of heavy interference.

**Comment by staowyze** [comment]
I didn't see how the state machine could get out of `TURN_STATE_FAILED`, do I miss anything?
https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/206eb882130a6d9cfa6bf401d315accf34cc0df0/src/source/Ice/TurnConnection.c#L1054

**Comment by chehefen** [comment]
hi @staowyze, TURN_STATE_FAILED is a terminal state. You can try to adjust the state timeout here: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/206eb882130a6d9cfa6bf401d315accf34cc0df0/src/source/Ice/TurnConnection.h#L23-L28 to have more tolerance. We will add capability to adjust these timeouts later.

**Comment by staowyze** [comment]
thanks for the information.

**Comment by chehefen** [comment]
Hi @staowyze, do you think having non-blocking TCP turn connection establishment helps with this issue? #197
